xdsl/transforms/csl_wrapper_to_csl.py

```diff
@@ -227,6 +227,76 @@ def lower_program_module(
         return program_module
 
 
+@dataclass(frozen=True)
+class LowerImport(RewritePattern):
+    """
+    Replace the `csl_wrapper.import` with the equivalent `csl.import`.
+
+    Hoist the import and all ops it depends on to the module scope (as is required by CSL)
+    """
+
+    def _get_csl_mod(self, op: Operation) -> csl.CslModuleOp:
+        """
+        Find the parent `csl.module` of the current op
+        """
+
+        if isinstance(op, csl.CslModuleOp):
+            return op
+        assert (parent := op.parent_op()) is not None
+        return self._get_csl_mod(parent)
+
+    def _collect_ops(self, op: Operation, ops: list[Operation]) -> list[Operation]:
+        """
+        Detach the op from its current location and store it in the list
+
+        Do this recursively for all operands of each operation.
+
+        NOTE: This op's dependencies are added to the list first to preserve
+              the order in which they get added back to the module.
+        """
+        op.detach()
+        for operand in op.operands:
+            owner = operand.owner
+            assert isinstance(owner, Operation)
+            self._collect_ops(owner, ops)
+        ops.append(op)
+        return ops
+
+    def _make_import_struct(self, import_op: csl_wrapper.ImportOp):
+        """
+        Create the struct to be passed to `@import_module`
+
+        Handles the case where multiple structs need to be concatinated before
+        being imported (this is indicated by an empty field name in
+        `csl_wrapper.import`). All required intermediate structs get returned
+        as a list.
+
+        Last struct in the list is to be used in `csl.import`
+        """
+
+        fields = list[tuple[str, SSAValue | Operation]]()
+        structs_to_concat = list[SSAValue | Operation]()
+        for fname, op in zip(import_op.fields, import_op.ops):
+            if fname.data != "":
+                fields.append((fname.data, op))
+            else:
+                structs_to_concat.append(op)
+        out_structs: list[Operation] = [csl.ConstStructOp(*fields)]
+        for s in structs_to_concat:
+            out_structs.append(csl.ConcatStructOp(out_structs[-1], s))
+        return out_structs
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: csl_wrapper.ImportOp, rewriter: PatternRewriter, /):
+        csl_mod = self._get_csl_mod(op)
+        ops = self._collect_ops(op, [])
```

Comment:
This seems to be running a series of `detach()` ops on the use-def chain. If you're concerned about leaving unused ops in the code, it might be easier to let the `dce`/`cse` pass handle the cleanup, and we might not need this method.

Reply:
This method moves the import statement to the module scope along with all of its dependencies. It does `detach() -> insert_op()` to preserve the links between all the operations as they are being moved . If there is an existing function that does this I'd love to use that instead.

Reply:
Agreed. I think placing all ops inside the loop nest is a good strategy, but any existing pass that does hoisting/loop-invariant-code-motion would help here. My thinking is that we do not currently analyse whether any of the instructions depend on the loop vars, but this may not be allowed in the first place so it's prob ok. Theoretically, the hoisting could always go in a small separate pass, but not strictly necessary.